File: backend/app/services/websocket_service.py
"""
WebSocket Service for Real-time Updates

This service handles WebSocket connections and broadcasts for the hybrid refresh strategy.
It categorizes events by priority and manages client connections.
"""

from flask_socketio import SocketIO, emit, join_room, leave_room
from flask import request
from typing import Dict, List, Optional, Any
from datetime import datetime
import json
import logging

logger = logging.getLogger(__name__)

class EventPriority:
    """Event priority levels for refresh strategy"""
    CRITICAL = "CRITICAL"  # Immediate refresh (0 delay)
    IMPORTANT = "IMPORTANT"  # 30-second delay with notification
    MINOR = "MINOR"  # 5-minute background sync

class EventCategory:
    """Event categories for different admin actions"""
    PAYMENT = "PAYMENT"
    ENROLLMENT = "ENROLLMENT"
    COURSE_UPDATE = "COURSE_UPDATE"
    USER_MANAGEMENT = "USER_MANAGEMENT"
    AVAILABILITY_CHANGE = "AVAILABILITY_CHANGE"
    SYSTEM_SETTINGS = "SYSTEM_SETTINGS"
    SESSION_CHANGE = "SESSION_CHANGE"
    PROFILE_UPDATE = "PROFILE_UPDATE"
    NOTIFICATION_NEW = "NOTIFICATION_NEW"
    NOTIFICATION_READ = "NOTIFICATION_READ"
    NOTIFICATION_DELETED = "NOTIFICATION_DELETED"

class WebSocketService:
    def __init__(self, socketio: SocketIO = None):
        self.socketio = socketio
        self.connected_clients: Dict[str, Dict] = {}
        self.event_queue: List[Dict] = []
        self.priority_mapping = {
            EventCategory.PAYMENT: EventPriority.CRITICAL,
            EventCategory.ENROLLMENT: EventPriority.CRITICAL,
            EventCategory.SESSION_CHANGE: EventPriority.CRITICAL,
            EventCategory.COURSE_UPDATE: EventPriority.IMPORTANT,
            EventCategory.AVAILABILITY_CHANGE: EventPriority.IMPORTANT,
            EventCategory.USER_MANAGEMENT: EventPriority.IMPORTANT,
            EventCategory.SYSTEM_SETTINGS: EventPriority.IMPORTANT,
            EventCategory.PROFILE_UPDATE: EventPriority.MINOR,
        }
    
    def init_app(self, socketio: SocketIO):
        """Initialize WebSocket service with SocketIO instance"""
        self.socketio = socketio
        self._register_handlers()
    
    def _register_handlers(self):
        """Register WebSocket event handlers"""
        if not self.socketio:
            return

        # Note: connect/disconnect handlers are already handled by chat_socket_service
        # We'll add our notification-specific handlers here
        @self.socketio.on('subscribe_to_notifications')
        def handle_notification_subscription():
            """Handle subscription to user-specific notifications"""
            try:
                from flask import session
                user_id = session.get('user_id')

                if user_id:
                    room_name = f"user_{user_id}"
                    join_room(room_name)
                    emit('notification_subscription_confirmed', {
                        'status': 'subscribed',
                        'room': room_name
                    })
                    logger.info(f"User {user_id} subscribed to notifications")
                else:
                    emit('error', {'message': 'Authentication required for notifications'})
            except Exception as e:
                logger.error(f"Error in notification subscription: {e}")
                emit('error', {'message': 'Failed to subscribe to notifications'})
    
    def _get_user_from_request(self) -> Optional[Dict]:
        """Extract user information from request context"""
        try:
            from flask_jwt_extended import get_jwt_identity, verify_jwt_in_request
            from app.models import User
            
            # Try to get user from JWT token
            verify_jwt_in_request(optional=True)
            user_id = get_jwt_identity()
            
            if user_id:
                user = User.query.get(user_id)
                if user:
                    return {
                        'id': user.id,
                        'account_type': user.account_type,
                        'email': user.email
                    }
        except Exception:
            # Try to get token from connection auth data
            if hasattr(request, 'event') and hasattr(request.event, 'get'):
                token = request.event.get('auth', {}).get('token')
                if token:
                    try:
                        from flask_jwt_extended import decode_token
                        from app.models import User
                        
                        decoded = decode_token(token)
                        user_id = decoded.get('sub')
                        if user_id:
                            user = User.query.get(user_id)
                            if user:
                                return {
                                    'id': user.id,
                                    'account_type': user.account_type,
                                    'email': user.email
                                }
                    except Exception:
                        pass
            
        return None
    
    def broadcast_admin_event(self, 
                            event_category: str,
                            event_data: Dict[str, Any],
                            affected_entities: Optional[List[Dict]] = None,
                            custom_priority: Optional[str] = None) -> bool:
        """
        Broadcast an admin event to relevant clients
        
        Args:
            event_category: Category of the event (from EventCategory)
            event_data: Data associated with the event
            affected_entities: List of entities affected (e.g., [{'type': 'course', 'id': 'course_123'}])
            custom_priority: Override default priority for this event
        
        Returns:
            bool: True if broadcast successful
        """
        try:
            if not self.socketio:
                logger.warning("WebSocket service not initialized")
                return False
            
            # Determine priority
            priority = custom_priority or self.priority_mapping.get(
                event_category, 
                EventPriority.MINOR
            )
            
            # Create event payload
            event_payload = {
                'event_id': f"evt_{int(datetime.utcnow().timestamp())}",
                'category': event_category,
                'priority': priority,
                'timestamp': datetime.utcnow().isoformat(),
                'data': event_data,
                'affected_entities': affected_entities or []
            }
            
            # Queue event for processing
            self.event_queue.append(event_payload)
            
            # Determine broadcast strategy based on priority
            if priority == EventPriority.CRITICAL:
                # Immediate broadcast to all users
                self._broadcast_immediate(event_payload)
            elif priority == EventPriority.IMPORTANT:
                # Broadcast with notification
                self._broadcast_with_notification(event_payload)
            else:
                # Queue for background sync
                self._queue_for_background_sync(event_payload)
            
            return True
            
        except Exception as e:
            logger.error(f"Error broadcasting admin event: {str(e)}")
            return False
    
    def _broadcast_immediate(self, event_payload: Dict):
        """Broadcast event immediately to all relevant clients"""
        try:
            # Broadcast to all connected clients
            self.socketio.emit('admin_update', event_payload)
            
            # Also broadcast to specific entity rooms if applicable
            for entity in event_payload.get('affected_entities', []):
                room_name = f"{entity['type']}_{entity['id']}"
                self.socketio.emit('entity_update', event_payload, room=room_name)
            
            logger.info(f"Critical event broadcast: {event_payload['event_id']}")
            
        except Exception as e:
            logger.error(f"Error in immediate broadcast: {str(e)}")
    
    def _broadcast_with_notification(self, event_payload: Dict):
        """Broadcast event with user notification"""
        try:
            # Add notification flag to payload
            event_payload['show_notification'] = True
            event_payload['notification_message'] = self._get_notification_message(event_payload)
            
            # Broadcast to all connected clients
            self.socketio.emit('admin_update_pending', event_payload)
            
            logger.info(f"Important event queued with notification: {event_payload['event_id']}")
            
        except Exception as e:
            logger.error(f"Error in notification broadcast: {str(e)}")
    
    def _queue_for_background_sync(self, event_payload: Dict):
        """Queue event for background synchronization"""
        try:
            # Add to background sync queue
            event_payload['sync_type'] = 'background'
            
            # Broadcast as low-priority update
            self.socketio.emit('background_sync_available', event_payload)
            
            logger.info(f"Minor event queued for background sync: {event_payload['event_id']}")
            
        except Exception as e:
            logger.error(f"Error in background sync queue: {str(e)}")
    
    def _get_notification_message(self, event_payload: Dict) -> str:
        """Generate user-friendly notification message based on event"""
        category = event_payload.get('category')
        messages = {
            EventCategory.COURSE_UPDATE: "Course information has been updated",
            EventCategory.ENROLLMENT: "Enrollment changes detected",
            EventCategory.SESSION_CHANGE: "Session schedule has been modified",
            EventCategory.AVAILABILITY_CHANGE: "Tutor availability has changed",
            EventCategory.USER_MANAGEMENT: "User information has been updated",
            EventCategory.SYSTEM_SETTINGS: "System settings have been modified",
            EventCategory.PAYMENT: "Payment information updated",
            EventCategory.PROFILE_UPDATE: "Profile information has been updated"
        }
        return messages.get(category, "Updates are available")
    
    def get_connected_clients_count(self) -> int:
        """Get count of currently connected clients"""
        return len(self.connected_clients)
    
    def get_event_queue_size(self) -> int:
        """Get size of current event queue"""
        return len(self.event_queue)
    
    def clear_event_queue(self):
        """Clear the event queue"""
        self.event_queue.clear()

    def broadcast_notification(self, user_id: str, notification_dict: dict, notification_type: str = "new") -> bool:
        """Send real-time notification to specific user"""
        try:
            if not self.socketio:
                logger.warning("WebSocket service not initialized")
                return False

            room_name = f"user_{user_id}"
            notification_data = {
                'type': notification_type,
                'notification': notification_dict,
                'timestamp': datetime.utcnow().isoformat(),
                'event_id': f"notif_{int(datetime.utcnow().timestamp())}"
            }

            self.socketio.emit('notification_event', notification_data, room=room_name)
            logger.info(f"Notification broadcast to user {user_id}: {notification_type}")
            return True

        except Exception as e:
            logger.error(f"Error broadcasting notification: {e}")
            return False

    def broadcast_unread_count_update(self, user_id: str, unread_count: int) -> bool:
        """Broadcast updated unread count to user"""
        try:
            if not self.socketio:
                logger.warning("WebSocket service not initialized")
                return False

            room_name = f"user_{user_id}"
            count_data = {
                'type': 'unread_count_update',
                'unread_count': unread_count,
                'timestamp': datetime.utcnow().isoformat()
            }

            self.socketio.emit('notification_count_update', count_data, room=room_name)
            logger.info(f"Unread count update broadcast to user {user_id}: {unread_count}")
            return True

        except Exception as e:
            logger.error(f"Error broadcasting unread count: {e}")
            return False

# Create global instance
websocket_service = WebSocketService()

File: frontend/src/services/websocketService.js
/**
 * WebSocket Service for Real-time Updates
 * 
 * This service manages WebSocket connections for the hybrid refresh strategy,
 * handling admin events and coordinating page refreshes based on priority.
 */

import io from 'socket.io-client';
import { refreshManager } from './refreshManager';
import smartUpdateService from './smartUpdateService';

class WebSocketService {
  constructor() {
    this.socket = null;
    this.isConnected = false;
    this.reconnectAttempts = 0;
    this.maxReconnectAttempts = 5;
    this.reconnectDelay = 1000;
    this.eventHandlers = new Map();
    this.subscriptions = new Set();
    this.config = {
      url: process.env.REACT_APP_WS_URL || process.env.REACT_APP_API_URL || 'http://localhost:80',
      transports: ['websocket', 'polling'],
      autoConnect: false,
      reconnection: true,
      reconnectionAttempts: 5,
      reconnectionDelay: 1000,
      reconnectionDelayMax: 5000,
      timeout: 20000
    };
  }

  /**
   * Initialize WebSocket connection
   */
  connect() {
    if (this.socket && this.isConnected) {
      console.log('WebSocket already connected');
      return;
    }

    try {
      // Get auth token for authentication
      const token = sessionStorage.getItem('authToken');
      
      // Create socket connection with auth
      this.socket = io(this.config.url, {
        ...this.config,
        auth: {
          token: token
        },
        query: {
          token: token
        }
      });

      this._setupEventHandlers();
      this.socket.connect();
      
    } catch (error) {
      console.error('Failed to initialize WebSocket:', error);
      this._fallbackToPolling();
    }
  }

  /**
   * Disconnect WebSocket
   */
  disconnect() {
    if (this.socket) {
      this.socket.disconnect();
      this.socket = null;
      this.isConnected = false;
      this.subscriptions.clear();
    }
  }

  /**
   * Setup WebSocket event handlers
   */
  _setupEventHandlers() {
    // Connection established
    this.socket.on('connect', () => {
      console.log('WebSocket connected');
      this.isConnected = true;
      this.reconnectAttempts = 0;
      
      // Re-subscribe to previous subscriptions
      this.subscriptions.forEach(subscription => {
        this.socket.emit('subscribe_to_updates', subscription);
      });
      
      // Notify refresh manager
      refreshManager.onWebSocketConnected();

      // Stop smart polling since WebSocket is connected
      smartUpdateService.stop();
    });

    // Connection lost
    this.socket.on('disconnect', (reason) => {
      console.log('WebSocket disconnected:', reason);
      this.isConnected = false;

      // Start smart polling as fallback
      this._fallbackToPolling();

      if (reason === 'io server disconnect') {
        // Server initiated disconnect, attempt reconnect
        setTimeout(() => this.connect(), this.reconnectDelay);
      }
    });

    // Connection error
    this.socket.on('connect_error', (error) => {
      console.error('WebSocket connection error:', error);
      this.reconnectAttempts++;
      
      if (this.reconnectAttempts >= this.maxReconnectAttempts) {
        console.log('Max reconnection attempts reached, falling back to polling');
        this._fallbackToPolling();
      }
    });

    // Admin update events (CRITICAL priority)
    this.socket.on('admin_update', (event) => {
      console.log('Received admin update:', event);
      this._handleAdminUpdate(event);
    });

    // Admin pending update events (IMPORTANT priority)
    this.socket.on('admin_update_pending', (event) => {
      console.log('Received pending admin update:', event);
      this._handlePendingUpdate(event);
    });

    // Background sync events (MINOR priority)
    this.socket.on('background_sync_available', (event) => {
      console.log('Background sync available:', event);
      this._handleBackgroundSync(event);
    });

    // Entity-specific updates
    this.socket.on('entity_update', (event) => {
      console.log('Received entity update:', event);
      this._handleEntityUpdate(event);
    });

    // Ping/pong for connection health
    this.socket.on('pong', (data) => {
      console.debug('Received pong:', data);
    });
  }

  /**
   * Handle critical admin updates (immediate refresh)
   */
  _handleAdminUpdate(event) {
    const { category, data, affected_entities, priority } = event;
    
    // Pass to refresh manager for immediate processing
    refreshManager.handleAdminEvent({
      type: 'ADMIN_UPDATE',
      priority: priority || 'CRITICAL',
      category,
      data,
      affectedEntities: affected_entities,
      timestamp: new Date().toISOString()
    });
  }

  /**
   * Handle pending updates with notification
   */
  _handlePendingUpdate(event) {
    const { category, data, affected_entities, notification_message } = event;
    
    // Pass to refresh manager with notification
    refreshManager.handleAdminEvent({
      type: 'ADMIN_UPDATE_PENDING',
      priority: 'IMPORTANT',
      category,
      data,
      affectedEntities: affected_entities,
      notificationMessage: notification_message,
      timestamp: new Date().toISOString()
    });
  }

  /**
   * Handle background sync events
   */
  _handleBackgroundSync(event) {
    const { category, data, affected_entities } = event;
    
    // Pass to refresh manager for background processing
    refreshManager.handleAdminEvent({
      type: 'BACKGROUND_SYNC',
      priority: 'MINOR',
      category,
      data,
      affectedEntities: affected_entities,
      timestamp: new Date().toISOString()
    });
  }

  /**
   * Handle entity-specific updates
   */
  _handleEntityUpdate(event) {
    const { category, data, affected_entities } = event;
    
    // Check if current page is affected
    const currentPath = window.location.pathname;
    const isAffected = this._isPageAffected(currentPath, affected_entities);
    
    if (isAffected) {
      refreshManager.handleEntityUpdate({
        category,
        data,
        affectedEntities: affected_entities,
        timestamp: new Date().toISOString()
      });
    }
  }

  /**
   * Check if current page is affected by the entities
   */
  _isPageAffected(currentPath, affectedEntities) {
    if (!affectedEntities || affectedEntities.length === 0) {
      return true; // If no specific entities, assume page is affected
    }

    for (const entity of affectedEntities) {
      const { type, id } = entity;
      
      // Check if current URL contains entity ID
      if (currentPath.includes(id)) {
        return true;
      }
      
      // Check specific page patterns
      if (type === 'course' && currentPath.includes('/course')) {
        return true;
      }
      if (type === 'user' && (currentPath.includes('/admin') || currentPath.includes('/profile'))) {
        return true;
      }
      if (type === 'session' && currentPath.includes('/session')) {
        return true;
      }
    }
    
    return false;
  }

  /**
   * Subscribe to updates for specific entity
   */
  subscribeToEntity(entityType, entityId) {
    const subscription = { entity_type: entityType, entity_id: entityId };
    
    this.subscriptions.add(subscription);
    
    if (this.socket && this.isConnected) {
      this.socket.emit('subscribe_to_updates', subscription);
    }
  }

  /**
   * Send ping to check connection health
   */
  ping() {
    if (this.socket && this.isConnected) {
      this.socket.emit('ping');
    }
  }

  /**
   * Fallback to smart polling when WebSocket fails
   */
  _fallbackToPolling() {
    console.log('WebSocket failed, falling back to smart polling mode');

    // Setup update handler for smart polling service
    smartUpdateService.onUpdate((updates) => {
      if (updates && updates.length > 0) {
        updates.forEach(update => {
          this._handleAdminUpdate(update);
        });
      }
    });

    // Start the smart polling service
    smartUpdateService.start();
  }

  /**
   * Clean up resources
   */
  destroy() {
    this.disconnect();

    // Stop smart polling service
    smartUpdateService.stop();

    this.eventHandlers.clear();
    this.subscriptions.clear();
  }
}

// Create and export singleton instance
const websocketService = new WebSocketService();
export default websocketService;